### PR TITLE
API-15279 F: Combine API-12976 and API-14307 functionality into one branch

### PR DESCRIPTION
**Pull request by mmccollum-va**
- Merge of efforts for `API-12976` without custom deserializers and `API-14307`
- Spring Caching elements removed from `API-12976`

### CODE REVIEW

**Comment by mmccollum-va** [comment]
@giangnguyen2 are you still seeing error mentioned above? Not able to reproduce it on V0 or V1 overlay upload endpoints.

**Comment by giangnguyen2** [comment]
> 654GF



> 



> @giangnguyen2 are you still seeing error mentioned above? Not able to reproduce it on V0 or V1 overlay upload endpoints.

Yes I still see the error. I put cms overlay from production to my local database and post a cms payload and get the error.

File: facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/DetailedService.java [hunk 1]
```diff
@@ -66,6 +68,7 @@ public class DetailedService implements CanBeEmpty {
 
   @JsonIgnore
   @Schema(description = "Deprecated until further notice.", example = "null", nullable = true)
+  @JsonAlias("description_facility")
```
[the remark this answers is not in the document]
Comment (mmccollum-va) [reply facilities-api/src/main/java/gov/va/api/lighthouse/facilities/api/v1/DetailedService.java]: Deprecated field `description_facility` removed from V1 and Datamart versions of `DetailedService`

**Comment by as-va** [comment]
I was able to reproduce @giangnguyen2's error report. It seems to be ok to add a new instance of the service to a facility that did not already have it, but posting a service to a facility that already had it prior to this update is generating the primary key violation.

**Comment by mmccollum-va** [comment]
Error observed when uploading an existing overlay for facility has been addressed; switch to utilize `serviceId` had enabled both capitalized and lowercase versions of a service to appear in `FacilityEntity` services collection.

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/FacilitiesCollector.java [hunk 2]
```diff
@@ -296,21 +296,23 @@ private void updateServicesFromCmsOverlay(List<DatamartFacility> datamartFacilit
...
-        datamartFacility.attributes().services().health(facilityHealthServiceList);
-      }
-    }
+
+    datamartFacilities.parallelStream()
+        .filter(df -> facilityCovid19Services.containsKey(df.id()))
+        .forEach(
+            df -> {
+              Set<HealthService> facilityHealthServices = new HashSet<>();
+              facilityHealthServices.add(facilityCovid19Services.get(df.id()));
```
Comment (giangnguyen2) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/FacilitiesCollector.java:305]: `parallelStream` might cause unexpected side effect as `HashSet` is not thread safe

Comment (mmccollum-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/collector/FacilitiesCollector.java:305]: Updated to utilize `stream()`

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java [hunk 3]
```diff
@@ -75,4 +120,221 @@ protected List<DatamartDetailedService> getOverlayDetailedServices(@NonNull Stri
...
+      if (facility.attributes().services().benefits() != null) {
+        facilityBenefitsServices.addAll(
+            facility.attributes().services().benefits().stream()
+                .filter(bs -> !disabledCmsServiceIds.contains(bs.serviceId()))
+                .collect(Collectors.toList()));
+      }
+      List<DatamartFacility.BenefitsService> facilityBenefitsServiceList =
+          new ArrayList<>(facilityBenefitsServices);
+      Collections.sort(facilityBenefitsServiceList);
+      facility.attributes().services().benefits(facilityBenefitsServiceList);
```
Comment (giangnguyen2) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:309]: Would this cause runtime error if `facility.attributes().services().benefits() == null`? Similar questions for health and other service list below

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java [hunk 4]
```diff
@@ -75,4 +120,221 @@ protected List<DatamartDetailedService> getOverlayDetailedServices(@NonNull Stri
...
+          (overlay.detailedServices() != null)
+              ? overlay.detailedServices().parallelStream()
+                  .filter(dds -> !dds.active())
+                  .map(dds -> dds.serviceInfo().serviceId())
+                  .collect(Collectors.toList())
+              : Collections.emptyList();
+
+      // Update facility benefits services
+      if (facility.attributes().services().benefits() != null) {
+        facilityBenefitsServices.addAll(
```
Comment (giangnguyen2) [inline facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:301]: When I pick a facility with the initial benefit service list empty, and post a cms payload for service id = `applyingForBenefits`  and `active flag = true`, the `facilityBenefitsServices` will be empty with the below filter condition and so the service is not saved into `facility.attributes().services().benefits` list

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java (hunk 3 above)
Comment (mmccollum-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:309]: Tested using `vha_402HL`, which initially contains no benefits, health, or other services. No `RuntimeException` was observed when overlay containing `Covid19Vaccine` detailed service was uploaded. `Covid19Vaccine` service appears correctly for `vha_402HL` facility services list and overlay detailed services.

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java (hunk 4 above)
Comment (mmccollum-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:301]: Tested using `vha_402HL`, which initially contains no benefits services. Able to successfully upload following overlay containing benefits service:
```
{
    "operating_status": {
        "code": "NORMAL",
        "additional_info": "Masks are required when in the facility."
    },
    "detailed_services": [
        {
            "name": "ApplyingForBenefits",
            "service_id": "applyingForBenefits",
            "active": "true",
            "appointment_leadin": "Your VA health care team will contact you if you’re eligible to get a vaccine during this time. As the supply of vaccine increases, we'll work with our care teams to let Veterans know their options.",
            "appointment_phones": [
                {
                    "extension": null,
                    "label": "Benefits Special",
                    "number": "252-867-5309",
                    "type": "tel"
                }
            ],
            "online_scheduling_available": null,
            "referral_required": "true",
            "walk_ins_accepted": "false",
            "service_locations": null,
            "path": "https://www.va.gov/durham-health-care/programs/covid-19-vaccines/"
        }
    ]
}
```

- `ApplyingForBenefits` appears in list of services for facility as well as overlay detailed services
- What facility ID are you observing the reported behavior?

Comment (giangnguyen2) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:301]: I notice now that if I put service name = `ApplyingForBenefits` then it work normally but if I put service name = "applyingForBenefits" then it doesn't get add to the list of facility services

{
  "detailed_services":[
    {
      "name":"applyingForBenefits",
	  "serviceId": "applyingForBenefits",
      "active":true,
      "description_facility":"This is not null",
      "health_service_api_id":null,
      "appointment_leadin":"Your VA health care team will contact you if you...more text",
      "online_scheduling_available": "True",
      "path": "\/erie-health-care\/locations\/erie-va-medical-center\/covid-19-vaccines",
      "appointment_phones": [
        {
          "type": "tel",
          "label": "Main phone",
          "number": "555-555-1212",
          "extension": "123" 
        }
      ],
      "referral_required": "False",
      "service_locations": null,
      "walk_ins_accepted": "True"
    }
  ],
   "operating_status": {
      "code": "NOTICE",
      "additional_info": "Addition info on operating_status"
    }
}

Comment (mmccollum-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java:301]: Hmm, would expect lowercase service name to work as well:
```
    @JsonCreator
    public static BenefitsService fromString(String name) {
      return eBenefitsRegistrationAssistance.name().equalsIgnoreCase(name)
          ? eBenefitsRegistrationAssistance
          : valueOf(capitalize(name));
    }
```

File: facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java [hunk 5]
```diff
@@ -75,4 +115,199 @@ protected List<DatamartDetailedService> getOverlayDetailedServices(@NonNull Stri
...
+    final Set<DatamartFacility.HealthService> facilityHealthServices =
+        Collections.synchronizedSet(new HashSet<>());
+    final Set<DatamartFacility.OtherService> facilityOtherServices =
+        Collections.synchronizedSet(new HashSet<>());
+
+    if (!toSaveDetailedServices.isEmpty()) {
+      final Set<String> detailedServiceIds = Collections.synchronizedSet(new HashSet<>());
+      toSaveDetailedServices.parallelStream()
+          .forEach(
+              service -> {
```
[the remark this answers is not in the document]
Comment (as-va) [reply facilities/src/main/java/gov/va/api/lighthouse/facilities/BaseCmsOverlayController.java]: Broken out into [API-16823](https://vajira.max.gov/browse/API-16823)

**Review by as-va — APPROVED** [review]
Approving for V0 functionality. Follow up tickets have been created for V1 behavior.
[API-16823](https://vajira.max.gov/browse/API-16823)
[API-16983](https://vajira.max.gov/browse/API-16983)

**Pull request merged**